lib/elixir/lib/calendar.ex

```diff
@@ -413,13 +523,79 @@ defmodule Date do
   """
   @spec compare(Calendar.date, Calendar.date) :: :lt | :eq | :gt
   def compare(date1, date2) do
-    case {to_erl(date1), to_erl(date2)} do
-      {first, second} when first > second -> :gt
-      {first, second} when first < second -> :lt
-      _ -> :eq
+    if Calendar.compatible_calendars?(date1.calendar, date2.calendar) do
+      case {to_rata_die(date1), to_rata_die(date2)} do
+        {first, second} when first > second -> :gt
+        {first, second} when first < second -> :lt
+        _ -> :eq
+      end
+    else
+      raise ArgumentError, """
+      cannot compare #{inspect date1} with #{inspect date2}:
+      This comparison would be ambiguous as their calendars have incompatible day rollover moments.
+      Specify an exact time of day (using `DateTime`s) to resolve this ambiguity.
+      """
+    end
+  end
+
+  def convert(%{calendar: calendar, year: year, month: month, day: day} = date, calendar), do: {:ok, date}
+  def convert(%{calendar: _, year: _, month: _, day: _} = date, target_calendar) do
+    if Calendar.compatible_calendars?(date.calendar, target_calendar) do
+      result_date =
+        date
+        |> to_rata_die
+        |> from_rata_die(target_calendar)
+      {:ok, result_date}
+    else
+      {:error, :incompatible_calendars}
+    end
+  end
+
+  @spec convert!(Date.t, Calendar.calendar) :: Date.t
+  def convert!(date, calendar) do
+    case convert(date, calendar) do
+      {:ok, value} ->
+        value
+      {:error, reason} ->
+        raise ArgumentError, "cannot convert #{inspect date} to target calendar #{inspect calendar}, reason: #{inspect reason}"
+    end
+  end
+
+  @doc """
+  Calculates the difference between two dates,
+  in a full number of days, returning `{:ok, difference}`.
+
+  Note that only Date structs that follow the same or compatible calendars
+  can be compared this way.
+  If two calendars are not compatible,
+  `{:error, :imcompatible_calendars}` is returned.
+
+  ## Examples
+
+      iex> Date.diff(~D[2000-01-01], ~D[2000-01-03])
+      {:ok, 2}
+  """
+  @spec diff(Date.t, Date.t) :: integer
+  def diff(%{calendar: _, year: _, month: _, day: _} = date1, %{calendar: _, year: _, month: _, day: _} = date2) do
```

Comment:
This function needs to be implemented like this:

```
def diff(%Date{} = date1, %Date{} = date2)
```

We must exclusively match on the Date struct because calculating Date.diff for naive datetimes will always give the wrong result.